README.md
# fasttime
## Fast Utility Functions for Time Parsing and Conversion

This package provides very fast functions for parsing date and timestamp formats. See [RForge page](https://rforge.net/fasttime) and [R documentation](https://rforge.net/doc/packages/fasttime/00Index.html).

[![CRAN](https://rforge.net/do/cransvg/fasttime)](https://cran.r-project.org/package=fasttime)
[![RForge](https://rforge.net/do/versvg/fasttime)](https://RForge.net/fasttime)
[![check](https://github.com/s-u/fasttime/actions/workflows/check.yml/badge.svg)](https://github.com/s-u/fasttime/actions/workflows/check.yml)

To install the CRAN version, use simply

```
install.packages("fasttime")
```

For installation of the latest development version, use

```
install.packages("fasttime", repo="https://rforge.net")
```


NEWS
NEWS for fasttime package

1.1-0	2022-03-16
    o	add fixed= parameter allowing to parse timestamps
	without separators (#4)

    o	add fastDate() (#5)


1.0-3	2018-01-05
    o	use registered C symbol for faster look-up


0.1-2	2016-10-26
    o	fix fastPOSIXct() example to use GMT when generating test
	timestamp range, otherwise R may generate timestamps before
	the epoch in timezones west of GMT (#2)

    o	fix handling of year 2100 which is not a leap year. The
	supported date range in fastPOSIXct() is now 1970..2199.

    o	fix 32-bit integer overflow when computing year offset (#3)


0.1-1	2015-07-27
    o	properly generate NA for invalid entries


0.1-0	2011-09-09
    o	first public version providing fastPOSIXct()


src/dparse.c
#define USE_RINTERNALS 1

#include <Rinternals.h>
#include <stdlib.h>
#include <stdint.h>

#define DIGIT(X) ((X) >= '0' && (X) <= '9')

/* start of each month in days */
static const int cml[] = { 0, 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365 };

typedef int64_t time_int_t;

SEXP parse_date(SEXP str, SEXP sRequiredComp) {
    SEXP res;
    double *tsv;
    int warn_nas = 0;
    int required_components = Rf_asInteger(sRequiredComp);
    int n, i, comp, ts;
    if (TYPEOF(str) != STRSXP) Rf_error("invalid date vector");
    n = LENGTH(str);
    res = Rf_allocVector(REALSXP, n);
    tsv = REAL(res);
    for (i = 0; i < n; i++) {
    	const char *c = CHAR(STRING_ELT(str, i));
        comp = 0;
        ts = 0;
    	if (DIGIT(*c)) {
    	    int y = 0, m = 0, d = 0;
    	    while (DIGIT(*c)) {
    	        y = y * 10 + (*c - '0');
    	        c++;
    	    }
    	    if (y < 100) {
    	        y += 30;   
    	    } else {
    	        y -= 1970;   
    	    }
    	    /* we only support the range of 1901-2500 to cover
    	       unsigned int dates without getting into more leap year mess */
    	    if (y <= -70) {
	            tsv[i] = NA_REAL;
    	        warn_nas = 1;
	            continue;
    	    } else {
        		/* adjust for all leap years prior to the current one
        		 * Work with dates before 1970 first
        		 */
                if (y < 0) {
                    if ((y-2) % 4 == 0) {
                        ts--;
                    }
                    ts += ((time_int_t)((y - 1) / 4)) * (time_int_t) 1;
                    
                /* Then work with dates after 1970 */
                } else {
                    ts += ((time_int_t)((y + 1) / 4)) * (time_int_t) 1;
                    if (y > 130) {
                        /* compensate for 2100, 2200, 2300, 2500:
                         * NOT leap years */
                        if (y <= 230) {
                            ts--;
                        } else if (y <= 330) {
                            ts -= 2;
                        } else if (y <= 530) {
                            ts -= 3;
                        } else {
                            tsv[i] = NA_REAL;
                            warn_nas = 1;
                            continue;
                        }
                    }
                }
        		
        		ts += ((time_int_t) y) * ((time_int_t) 365);
        		comp++;
        		while (*c && !DIGIT(*c)) c++;
        		if (*c) {
        		    while (DIGIT(*c)) {
        		        m = m * 10 + (*c - '0');
        		        c++;
        		    }
        		    if (m > 0 && m < 13) {
            			ts += cml[m];
            			if (m > 2 && (y & 3) == 2 &&
                            y != 130 && y !=  230 && y != 330 && y != 530 /* 2100 again */) {
            			    ts++;
            			}
            			comp++;
            			while (*c && !DIGIT(*c)) c++;
            			if (*c) {
            			    while (DIGIT(*c)) { d = d * 10 + (*c - '0'); c++; }
            			    if (d > 1) ts += (d - 1) * 1;
            			    comp++;
            			}
        		    }
        		}
    	    }
    	}
    	tsv[i] = (comp >= required_components) ? ts : NA_REAL;
    }
    /* send a warning to the console if NAs were introduced */
    if (warn_nas == 1) {
        Rf_warning("NAs introduced for out of range dates.");
    }
    return res;
}


src/tparse.c
#define USE_RINTERNALS 1

#include <Rinternals.h>
#include <stdlib.h>
#include <stdint.h>
#include <stddef.h>
#include <string.h>

#define DIGIT(X) ((X) >= '0' && (X) <= '9')

/* start of each month in seconds */
static const int cml[] = { 0, 0, 2678400, 5097600, 7776000, 10368000, 13046400, 15638400,
			   18316800, 20995200, 23587200, 26265600, 28857600, 31536000 };

typedef int64_t time_int_t;

/* separated input version, argitrary length of each component */
static SEXP parse_ts_(SEXP str, SEXP sRequiredComp, int date) {
    SEXP res;
    double *tsv;
    int required_components = Rf_asInteger(sRequiredComp);
    size_t n, i;
    if (TYPEOF(str) != STRSXP) Rf_error("invalid timestamp vector");
    n = XLENGTH(str);
    res = Rf_allocVector(REALSXP, n);
    tsv = REAL(res);
    for (i = 0; i < n; i++) {
	const char *c = CHAR(STRING_ELT(str, i));
	int comp = 0;
	double ts = 0.0;
	if (DIGIT(*c)) {
	    int y = 0, m = 0, d = 0, h = 0, mi = 0;
	    while (DIGIT(*c)) { y = y * 10 + (*c - '0'); c++; }
	    if (y < 100) y += 2000;
	    y -= 1970;
	    /* we only support the range of 1970-2199 to cover
	       unsigned int POSIX time without getting into more leap year mess */
	    if (y < 0 || y >= 230 ) {
		tsv[i] = NA_REAL;
		continue;
	    } else {
		/* adjust for all leap years prior to the current one */
	      ts += ((time_int_t)((y + 1) / 4)) * (time_int_t) 86400;
		if (y > 130) /* 2100 is an exception - not a leap year */
		    ts -= 86400;
		ts += ((time_int_t) y) * ((time_int_t) 31536000);
		comp++;
		while (*c && !DIGIT(*c)) c++;
		if (*c) {
		    while (DIGIT(*c)) { m = m * 10 + (*c - '0'); c++; }
		    if (m > 0 && m < 13) {
			ts += cml[m];
			if (m > 2 && (y & 3) == 2 && y != 130 /* 2100 again */) ts += 86400;
			comp++;
			while (*c && !DIGIT(*c)) c++;
			if (*c) {
			    while (DIGIT(*c)) { d = d * 10 + (*c - '0'); c++; }
			    if (d > 1) ts += (d - 1) * 86400;
			    comp++;
			    while (*c && !DIGIT(*c)) c++;
			    if (*c) {
				while (DIGIT(*c)) { h = h * 10 + (*c - '0'); c++; }
				ts += h * 3600;
				comp++;
				while (*c && !DIGIT(*c)) c++;
				if (*c) {
				    while (DIGIT(*c)) { mi = mi * 10 + (*c - '0'); c++; }
				    ts += mi * 60;
				    comp++;
				    while (*c && !(DIGIT(*c) || *c == '.')) c++;
				    if (*c) {
					ts += atof(c);
					comp++;
				    }
				}
			    }
			}
		    }
		}
	    }
	}
	tsv[i] = (comp >= required_components) ? (date ? ts / 86400 : ts) : NA_REAL;
    }
    return res;
}

/* fixed version - limits components to <4/2><2><2><2><2><*>
   and does not need separators. Note that it will still skip
   separators if present */
static SEXP fparse_ts_(SEXP str, SEXP sRequiredComp, int date, int year_length) {
    SEXP res;
    double *tsv;
    int required_components = Rf_asInteger(sRequiredComp);
    size_t n, i;
    if (TYPEOF(str) != STRSXP) Rf_error("invalid timestamp vector");
    n = XLENGTH(str);
    res = Rf_allocVector(REALSXP, n);
    tsv = REAL(res);
    for (i = 0; i < n; i++) {
	const char *c = CHAR(STRING_ELT(str, i));
	int comp = 0;
	double ts = 0.0;
	if (DIGIT(*c)) {
	    int y = 0, m = 0, d = 0, h = 0, mi = 0, left = year_length;
	    while (left-- && DIGIT(*c)) { y = y * 10 + (*c - '0'); c++; }
	    if (y < 100) y += 2000;
	    y -= 1970;
	    /* we only support the range of 1970-2199 to cover
	       unsigned int POSIX time without getting into more leap year mess */
	    if (y < 0 || y >= 230 ) {
		tsv[i] = NA_REAL;
		continue;
	    } else {
		/* adjust for all leap years prior to the current one */
	      ts += ((time_int_t)((y + 1) / 4)) * (time_int_t) 86400;
		if (y > 130) /* 2100 is an exception - not a leap year */
		    ts -= 86400;
		ts += ((time_int_t) y) * ((time_int_t) 31536000);
		comp++;
		while (*c && !DIGIT(*c)) c++;
		if (*c) {
		    int monl = 2;
		    while (monl-- && DIGIT(*c)) { m = m * 10 + (*c - '0'); c++; }
		    if (m > 0 && m < 13) {
			ts += cml[m];
			if (m > 2 && (y & 3) == 2 && y != 130 /* 2100 again */) ts += 86400;
			comp++;
			while (*c && !DIGIT(*c)) c++;
			if (*c) {
			    int dayl = 2;
			    while (dayl-- && DIGIT(*c)) { d = d * 10 + (*c - '0'); c++; }
			    if (d > 1) ts += (d - 1) * 86400;
			    comp++;
			    while (*c && !DIGIT(*c)) c++;
			    if (*c) {
				int hleft = 2;
				while (hleft-- && DIGIT(*c)) { h = h * 10 + (*c - '0'); c++; }
				ts += h * 3600;
				comp++;
				while (*c && !DIGIT(*c)) c++;
				if (*c) {
				    int mileft = 2;
				    while (mileft-- && DIGIT(*c)) { mi = mi * 10 + (*c - '0'); c++; }
				    ts += mi * 60;
				    comp++;
				    while (*c && !(DIGIT(*c) || *c == '.')) c++;
				    if (*c) {
					ts += atof(c);
					comp++;
				    }
				}
			    }
			}
		    }
		}
	    }
	}
	tsv[i] = (comp >= required_components) ? (date ? ts / 86400 : ts) : NA_REAL;
    }
    return res;
}

/* -- API -- */

static SEXP sym_tzone;

SEXP parse_ts(SEXP str, SEXP sRequiredComp, SEXP sFixed, SEXP sTZ) {
    SEXP res, cls;
    int fixed = Rf_asInteger(sFixed);
    int date = 0;

    /* we (ab)use tz of "Date" to signal that the user wants a date ;) */
    if (TYPEOF(sTZ) == STRSXP &&
	LENGTH(sTZ) == 1 &&
	!strcmp(CHAR(STRING_ELT(sTZ, 0)), "Date"))
	date = 1;

    /* call the correct parser */
    res = (fixed < 1) ? parse_ts_(str, sRequiredComp, date) : fparse_ts_(str, sRequiredComp, date, fixed);

    /* non-string values will make us return the naked object */
    if (TYPEOF(sTZ) != STRSXP && sTZ != R_NilValue) return res;
    PROTECT(res);
    if (date)
	Rf_classgets(res, Rf_mkString("Date"));
    else {
	/* if TZ is a string, turn this into a POSIXct object */
	if (sTZ != R_NilValue) {
	    if (!sym_tzone)
		sym_tzone = Rf_install("tzone");
	    Rf_setAttrib(res, sym_tzone, sTZ);
	}
	cls = PROTECT(allocVector(STRSXP, 2));
	SET_STRING_ELT(cls, 0, Rf_mkChar("POSIXct"));
	SET_STRING_ELT(cls, 1, Rf_mkChar("POSIXt"));    
	Rf_classgets(res, cls);
	UNPROTECT(1);
    }
    UNPROTECT(1);
    return res;
}



/* registration */

#include <R_ext/Rdynload.h>

static const R_CallMethodDef CallEntries[] = {
    {"parse_ts", (DL_FUNC) &parse_ts, 4},
    {NULL,       NULL,                0}
};

void R_init_fasttime(DllInfo *dll)
{
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}
